### ISSUE TITLE
Support for Other Platforms Like Linux and macOS

### ISSUE THREAD

**Issue by mehmetakiftutuncu** [fr31/spotifylyrics#19]
Hi there!

I'd like to try your app but I don't have experience with Python so I don't exactly know how to run it from the sources. Is it possible to either write a document on how to do this or simply providing executables for other platforms?

I am running OS X El Capitan by the way.

Thanks.

**Comment by fr31** [comment]
I'm not sure if it even works on os x tbh. It might, if the linux method of getting it to work also works on os x, but I don't have an os x machine to test it on.

_If_ it does work however, the directions should be similar to the linux directions but I'll try to make it more clear.

1 . Download and install [latest python](https://www.python.org/downloads/release/python-352/).
2 . Open up the terminal and install the dependencies.

`pip3 install bs4 requests xmltodict pyqt5`
(if that doesn't work try putting "sudo" in front of it)

3 . Download the "master" branch on github. On the main page click the green "Clone or Download" button and then click "Download ZIP"

4 . Extract zip, open terminal, change directories to extracted zip using "cd /directoryhere" and run with "python3 SpotifyLyrics.pyw" in the terminal.


These may not be exactly right as I don't have a mac, but hopefully they're close enough, if someone does have it working on os x though, feel free to chime in with better directions.

If it _doesn't_ work, well, not much I can really do my self. But if anyone who knows python and has os  x, you can make it work  _hint hint nudge nudge_

**Comment by mehmetakiftutuncu** [comment]
It cannot find `dbus` module. I'm getting:

```
Traceback (most recent call last):
  File "SpotifyLyrics.pyw", line 2, in <module>
    import backend
  File "/Users/akif/Downloads/spotifylyrics-master/backend.py", line 15, in <module>
    import dbus
ImportError: No module named 'dbus'
```

I tried installing `homebrew/python/python-dbus` package via Homebrew but I doubt it is for Python 3. I Googled how to install `dbus` on a Mac but didn't find much. I'll look into it when I have more time.

Thanks for your help. :)

**Comment by fr31** [comment]
`pip3 install dbus`

perhaps?

If not, the use of dbus is fairly recent. Maybe give the 1.09 zip a try. It doesn't use dbus at all.

https://github.com/fr31/spotifylyrics/archive/1.09.zip

**Comment by mehmetakiftutuncu** [comment]
`pip3` couldn't find `dbus` saying `Collecting dbus
  Could not find a version that satisfies the requirement dbus (from versions: )
No matching distribution found for dbus`. That was the first thing I've tried actually. :) 1.09 did run but it didn't detect the song that was playing in Spotify. I restarted Spotify, it didn't help. I had following output in the terminal by the way:

```
libpng warning: iCCP: known incorrect sRGB profile
/bin/bash: xwininfo: command not found
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/local/Cellar/python3/3.5.2_1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/threading.py", line 914, in _bootstrap_inner
    self.run()
  File "/usr/local/Cellar/python3/3.5.2_1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/threading.py", line 862, in run
    self._target(*self._args, **self._kwargs)
  File "SpotifyLyrics.pyw", line 232, in lyrics_thread
    songname = backend.getwindowtitle()
  File "/Users/akif/Downloads/spotifylyrics-1.09/backend.py", line 159, in getwindowtitle
    windows = subprocess.check_output(["/bin/bash", "-c", command]).decode("utf-8")
  File "/usr/local/Cellar/python3/3.5.2_1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/subprocess.py", line 626, in check_output
    **kwargs).stdout
  File "/usr/local/Cellar/python3/3.5.2_1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/subprocess.py", line 708, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['/bin/bash', '-c', 'xwininfo -tree -root']' returned non-zero exit status 127
```

**Comment by fr31** [comment]
Hm. It looks like it's not gonna run properly on os x in it's current form. Sorry.

If anyone who knows python, has a mac, and wants to get it working, here are some resources which may be helpful for getting it working:

https://developer.spotify.com/applescript-api/
http://stackoverflow.com/questions/28815863/how-to-get-active-window-title-using-python-in-mac-os

**Comment by archagy** [comment]
I have a mac. I will check it, in my free time at home.

**Comment by Dennisbonke** [comment]
Can confirm that linux from source works. Tested on Kali Linux 2.0 Rolling, kernel 4.6.0. However, on Kali, replace the pip install commands with apt-get install python3-bs4 python3-requests python3-xmltodict python3-pyqt5, furthermore, xwininfo comes preinstalled on Kali, which package provides it is to me unknown as of yet. Thank you for this amazing replacement for the missing lyrics feature on Spotify!

**Comment by alexfornuto** [comment]
I've got it running on OS X, just not "working" You need [brew](brew.sh/)

```
brew install d-bus dbus-glib
pip3 install dbus-python
```
You also seem to need to manually create and start the dbus service. There are several methods floating out there. I belive the one that worked for me was:

```
ln -sfv /usr/local/opt/d-bus/*.plist ~/Library/LaunchAgents
launchctl load ~/Library/LaunchAgents/org.freedesktop.dbus-session.plist
```

which I pulled from [here](http://blog.roderickmann.org/2015/03/using-dbus-on-os-x-yosemite/).

I was then able to run `python3 SpotifyLyrics.pyw`. The GUI opened, but it doesn't seem to see my running Spotify instance:

```
$ python3 SpotifyLyrics.pyw
libpng warning: iCCP: known incorrect sRGB profile
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/site-packages/dbus/bus.py", line 175, in activate_name_owner
    return self.get_name_owner(bus_name)
  File "/usr/local/lib/python3.5/site-packages/dbus/bus.py", line 361, in get_name_owner
    's', (bus_name,), **keywords)
  File "/usr/local/lib/python3.5/site-packages/dbus/connection.py", line 651, in call_blocking
    message, timeout)
dbus.exceptions.DBusException: org.freedesktop.DBus.Error.NameHasNoOwner: Could not get owner of name 'org.mpris.MediaPlayer2.spotify': no such name

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/Cellar/python3/3.5.2_1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/threading.py", line 914, in _bootstrap_inner
    self.run()
  File "/usr/local/Cellar/python3/3.5.2_1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/threading.py", line 862, in run
    self._target(*self._args, **self._kwargs)
  File "SpotifyLyrics.pyw", line 240, in lyrics_thread
    songname = backend.getwindowtitle()
  File "/Users/alex/repos/spotifylyrics/backend.py", line 191, in getwindowtitle
    spotifydbus = session.get_object("org.mpris.MediaPlayer2.spotify", "/org/mpris/MediaPlayer2")
  File "/usr/local/lib/python3.5/site-packages/dbus/bus.py", line 241, in get_object
    follow_name_owner_changes=follow_name_owner_changes)
  File "/usr/local/lib/python3.5/site-packages/dbus/proxies.py", line 248, in __init__
    self._named_service = conn.activate_name_owner(bus_name)
  File "/usr/local/lib/python3.5/site-packages/dbus/bus.py", line 180, in activate_name_owner
    self.start_service_by_name(bus_name)
  File "/usr/local/lib/python3.5/site-packages/dbus/bus.py", line 278, in start_service_by_name
    'su', (bus_name, flags)))
  File "/usr/local/lib/python3.5/site-packages/dbus/connection.py", line 651, in call_blocking
    message, timeout)
dbus.exceptions.DBusException: org.freedesktop.DBus.Error.ServiceUnknown: The name org.mpris.MediaPlayer2.spotify was not provided by any .service files
```

Meanwhile in the GUI, all I get is "Play a song in Spotify to fetch lyrics." Still, progress right?

**Comment by mehmetakiftutuncu** [comment]
@alexfornuto Looks like we had similar progress. Still no luck with running properly though. :)

**Comment by keryil** [comment]
I have created a pull request ( #23 ) with a hack I put together just now. It uses Spotify's AppleScript API to request the artist and song name.

**Comment by keryil** [comment]
Well, even with the hack, the sync doesn't work because that update loop that depends on detecting changes to the title of the Spotify window, which doesn't happen under OS/X. I will work on that.

**Comment by fr31** [comment]
Nice. Merged.

**Comment by leocaseiro** [comment]
Hi @keryil, thanks for helping with this issue.

I use the [amplify](https://github.com/ezig/amplify) with OSX and it seems pretty sync. Its source might help you with the sync.

That would be awesome having it running on OSX.

Thanks

**Comment by autMia** [comment]
I have basic functionality on macOS 10.12.1 after:

- Installing Python3 and PyQt5 through brew
- Installing requirements through pip3
- Running with python3

Unfortunately synced lyrics doesn't work as stated above.

**Comment by barsnick** [comment]
Regarding Linux, this should probably pull in all required packages (apart from spotify(-client)) on Fedora:
```
sudo dnf install python3 python3-dbus python3-qt5 python3-xmltodict python3-sip python3-beautifulsoup4 python3-requests xorg-x11-utils
```

This worked for me on Fedora 24. I couldn't test on Fedora 25, because the machine I have already updated is only available via X11 through ssh, and one of those python modules segfaults in that scenario. ;-)

I'll try to provide an RPM of this tool via my copr repository. Getting spotify-client on Linux to run is then everyone else's problem. ;-) rpmfusion no longer provide a nice LPF package (that's a package which helps you build your own package, without violating copyright).

**Comment by frostworx** [comment]
great work, thanks a lot!

**Comment by mirh** [comment]
Fixed with #54 ?

**Comment by fr31** [comment]
I'm gonna close this one since it works on most platforms now but if anyone has a specific issue with a certain platform feel free to open another issue.
